plaso/parsers/gcp_logging.py

```diff
@@ -0,0 +1,293 @@
+          protoPayload field.
+      event_attributes (dict): a dict representing event attributes to be added
+        to the event object.
+    """
+    request_attributes = [
+        'name', 'description', 'direction', 'member', 'targetTags', 'email',
+        'account_id'
+    ]
+    for attribute in request_attributes:
+      if attribute in request:
+        plaso_attribute = 'request_{0:s}'.format(attribute)
+        event_attributes[plaso_attribute] = request[attribute]
+
+    # Firewall specific attributes.
+    if 'sourceRanges' in request:
+      source_ranges = ', '.join(request['sourceRanges'])
+      event_attributes['source_ranges'] = source_ranges
+
+    if 'alloweds' in request:
+      for allowed in request['alloweds']:
+        attribute_name = 'allowed_{0:s}_ports'.format(allowed['IPProtocol'])
+        if 'ports' in allowed:
+          event_attributes[attribute_name] = allowed['ports']
+        else:
+          event_attributes[attribute_name] = 'all'
+
+    if 'denieds' in request:
+      for denied in request['denieds']:
+        attribute_name = 'denied_{0:s}_ports'.format(denied['IPProtocol'])
+        if 'ports' in denied:
+          event_attributes[attribute_name] = denied['ports']
+        else:
+          event_attributes[attribute_name] = 'all'
+
+    # Service account specific attributes
+    if 'service_account' in request:
+      service_account_name = request['service_account'].get('display_name')
+      event_attributes['service_account_display_name'] = service_account_name
+
+  def _ParseJSONPayload(self, json_payload, event_attributes):
+    """Extracts information from a json_payload.
+
+    Args:
+      json_payload (dict): the contents of a GCP jsonPayload field.
+      event_attributes (dict): a dict representing event attributes to be added
+        to the event object.
+    """
+    json_attributes = [
+        'event_type', 'event_subtype', 'container', 'filename', 'message'
+    ]
+    for attribute in json_attributes:
+      if attribute in json_payload:
+        event_attributes[attribute] = json_payload[attribute]
+
+    actor = json_payload.get('actor', {})
+    if actor:
+      if 'user' in actor:
+        event_attributes['user'] = actor['user']
+
+  def ParseFileObject(self, parser_mediator, file_object):
+    """Parses GCP logging saved to a file.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      file_object (dfvfs.FileIO): a file-like object.
+
+    Raises:
+      UnableToParseFile: when the file cannot be parsed.
+      ValueError: if the JSON file cannot be decoded.
+    """
+    if file_object.read(1) != b'[':
+      raise errors.UnableToParseFile(
+          'is not a valid JSON file, missing opening brace.')
+    file_object.seek(0, os.SEEK_SET)
+    try:
+      file_json = json.loads(file_object.read().decode('utf-8'))
```

Comment:
Yeah that's a good point especially as it will call read() on any file that passes the initial check. I will change the parser to use jsonl so that logs can be parsed one line at a time rather than having to read the whole file up front.